Installs the .la files in the install location

permcody opened the pull request:
This PR appears to work on Mac for installed MOOSE applications. However, there are as many unanswered questions as there are answered ones with this. First to get this working for our dynamic load case (MacOS only), one most set `DYLD_LIBRARY_PATH` to point to the multiapp's lib location. This is probably not something we want to do in the long run. I have no idea how this behaves yet on Linux. 

What I'm discovering though is that Mac may be automatically loading dependent libraries. I'm making this assumption based on the following error:
```
dlopen(/Users/permcj/projects/installed/lib/bison/libxfem-opt.0.dylib, 0x0001): Library not loaded: @rpath/libtensor_mechanics-opt.0.dylib
  Referenced from: <2DFB3AA0-1CE7-3564-A3FF-96F2369FCAC0> /Users/permcj/projects/installed/lib/bison/libxfem-opt.0.dylib
  Reason: tried: '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/libmesh-vtk/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/permcj/mambaforge3/envs/moose/lib/gcc/arm64-apple-darwin20.0.0/11.0.1/libtensor_mechanics-opt.0.dylib' (no such file), '/Users/permcj/projects/installed/lib/pronghorn/./libtensor_mechanics-opt.0.dylib' (no such file), '/usr/local/lib/libtensor_mechanics-opt.0.dylib' (no such file), '/usr/lib/libtensor_mechanics-opt.0.dylib' (no such file, not in dyld cache)
```

Note all of the paths being search - automatically. This is the second library that's "loaded" by MOOSE. The first being the one being searched for above.

**UPDATE (5/1):** With the latest patch this appears to be working on both Mac and Linux with in-tree and installed binaries. Also the double-free error on tear down is gone!

There were two major confounding factors that lead to what I believe was an implementation problem of this system _since its inception_. Namely, it doesn't appear that manually loading "inter-library dependencies" is necessary despite my long held belief. I believe that the primary issue was the lack of uniqueness in a specific symbol used for loading that was introduced to create our beautiful modules system. Namely the `Modules::registerAll` method is duplicated for each application even though the implementation in the body of that method is dependent based on the specific set of modules enabled. While this design error was successfully covered by compiling applications together (since our library names are unique), it appears to have been a primary cause of some of the longstanding difficulty of missing module registration for our dynamically opened libraries. With the latest patches, I'm successfully opening _just_ the top-level application library dependency and successfully running input files that include objects that are _not_ in the driving application. 

Fixes: With this PR, I'm deprecating `Modules::registerAll` in favor of `Modules::registerAllObjects<AppName>`. This will effectively give us back a unique entry point for each application. I'm also turning off the manually recursive library load logic. 

Intermediate Debugging and Rollout: For right now I'm going to continue to install the .la files, but if everything in this PR works out as planned, we won't need those anymore. The libraries themselves should contain the information needed to properly load the dependent libraries with out library archive files so that will come back out. Also, I'm going to add a Boolean to allow us to toggle the manually recursive loading just in case. Again, I don't expect we'll need this, but it maintains the logic we've had in there for the past ~8 years at least for now in case we need it in a pinch. That can also come out after sufficient testing/usage. Apps (at least the nuclear herd) will need to update to the new module registration method, but the old one can stick around for a while. Several months down the road we should be able to significantlly clean all of this loading logic up. We can probably cut the total lines in those routines by upwards of 70% since we only need to load a single library now. 

Testing: We still don't have a good way of testing all of this. I know some want to create new dummy applications that emulate some of our more complex herd animals, but I don't think that's effective. I've found through all of the implementation that being close (like I was 8 years ago) is not anywhere close to good enough. We need to understand why a specific combination isn't working with all the real-world messy code that each of these applications brings along. At the very least we should be using real applications to test even if we pick specific input files to use in testing. For my testing, I've been using real Pronghorn and BISON input files and using `FullSolveMultiapp` to just run the subapp input file. That's generally enough to give me a warm fuzzy that the subapp is being properly dynamically loaded and fully working. Running `simple_transient_diffusion` is not good enough. We need to know that downstream dependencies are being used and working properly. With the BISON/Pronghorn example, I found many input files that worked because many were using a subset of objects that Pronghorn new about. It wasn't until I stumbled upon files containing XFEM (a module not used by Pronghorn) that I started to unravel this whole mystery. So for testing we do need disjoint subsets of modules enabled and input files that exercise objects from the set difference. 

Should we merge this before we figure out all the testing?
I'm biased, but yes. We know that the halfway baked implementation is already working well for some of our users (notably the microreactor team at ANL), but that's because they are getting lucky with modules. This PR installs the .la files (but those normally won't be used, which also means the python script to patch them also won't be used). Again, we'll pull all that back out later. The main "merge me now" changes here are the two I started this update with (adding a new templated modules registration method - very low impact, and turning off recursive loading). The former is a new method so won't affect existing testing and the second feature can only affect dynamic loading.

permcody committed:
Install library archives (unmodified) for dynamic loading of binary installs

refs #12266

permcody committed:
Patching the .la files so dependency paths make sense

This PR is working on MacOS, but only if I also set DYLD_LIBRARY_PATH to point to the location
where the libraries of the Multiapp is located. I've tracked this down a bit and the issue is
that Mac is ignoring the fact that I'm explicitly loading the dependent librarie (might be my fault)
and it's only paying attention to @rpath for where to look for dependent libraries even though I'm
explicitly trying to load them. However, we normally won't have the @rpaths of one application
stored in another (e.g. pronghorn doesn't know about our installed location of bison).

Bottom line, DYLD_LIBRARY_PATHS works for at least library interdependencies. I have not investigated this
on Linux yet

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1462459/) on 8847b31 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24064/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1462473/) on 8847b31 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24064/coverage/xfem)

</details>



This comment will be updated on new commits.

permcody committed:
Makefile improvements

 - better dependencies
 - don't build test libraries when installing

closes #24079

permcody committed:
I don't think we need this logic

permcody committed:
Better dynamic library loading errors

- add lderror() to the dynamic load failure message
- print out the search paths instead of the loaded paths when dynamic loading fails

permcody committed:
WIP: Trying to not install la and not load dependencies

permcody committed:
Back to install library archives

We are going to reinstall library archives, but this time we are also going to decide on
dynamic library uniqueness just based on archive file name (not the full path). This will mean
that we'll only get one version of "heat condition" for instance instead of two copies just because
they are in different directories.

permcody committed:
We need to load dependencies by default. I may make this a user controlled param.

permcody committed:
We really should have a sane default prefix

lindsayad commented on framework/build.mk:
```diff
@@ -11,6 +11,10 @@ MOOSE_JOBS        ?= 8
 # Include variables defined by MOOSE configure if it's been run
 -include $(MOOSE_DIR)/conf_vars.mk
 
+# PREFIX should be set in conf_vars. If however the user didn't run configure we
+# won't have a prefix. We'll use the automake default then:
+PREFIX ?= '/usr/local'
```
I feel like I've been destroyed by defaults like this in the past

lindsayad replied:
But I guess this a default for where *you* are installing? Not a default for where to look for packages?

permcody replied:
`/usr/local` is the normal "automake" default and yes, this is for where you will install, not look for anything. However, since we don't require that users run "configure". We evaluate $PREFIX" in our Makefile and it's empty. So our "no configure" default is actually `/`. If you run "configure" without any arguments, you get an explicit PREFIX of `/usr/local`. To me, it's sane to make those two the same, which this will do.

permcody committed:
Adjusting error message in test

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1479259/) on de835c9 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 85.0%)

lindsayad commented on framework/app.mk:
```diff
@@ -431,6 +431,11 @@ else
   install_bin:
 endif
 
+#lib_install_targets = $(foreach lib,$(applibs),$(dir $(lib))install_lib_$(notdir $(lib)))
+#ifneq ($(app_test_LIB),)
+#	lib_install_targets += $(dir $(app_test_LIB))install_lib_$(notdir $(app_test_LIB))
+#endif
+
```
```suggestion
```

lindsayad commented on framework/app.mk:
```diff
@@ -469,18 +474,22 @@ $(copy_input_targets):
-	@$(eval libnames := $(foreach lib,$(applibs),$(shell grep "dlname='.*'" $(lib) 2>/dev/null | sed -E "s/dlname='(.*)'/\1/g")))
-	@$(eval libpaths := $(foreach lib,$(applibs),$(dir $(lib))$(shell grep "dlname='.*'" $(lib) 2>/dev/null | sed -E "s/dlname='(.*)'/\1/g")))
-	@for lib in $(libpaths); do $(call patch_relink,$(libdst),$$lib,$$(basename $$lib)); done
+#@$(eval libnames := $(foreach lib,$(applibs),$(shell grep "dlname='.*'" $(lib) 2>/dev/null | sed -E "s/dlname='(.*)'/\1/g")))
+#	@$(eval libpaths := $(foreach lib,$(applibs),$(dir $(lib))$(shell grep "dlname='.*'" $(lib) 2>/dev/null | sed -E "s/dlname='(.*)'/\1/g")))
+#	@for lib in $(libpaths); do $(call patch_relink,$(libdst),$$lib,$$(basename $$lib)); done
```
```suggestion
```

lindsayad commented on framework/scripts/patch_la.py:
```diff
@@ -0,0 +1,42 @@
+    shutil.copystat(filename, filename + '~tmp')
+    os.rename(filename + '~tmp', filename)
+
+if len(sys.argv) == 3:
```
Should we error if the length is something other than 3?

permcody replied:
Maybe, I'm following the exact same patter of our `get_repo_revision` script that we've used for more than a decade. It just doesn't do anything if it's not called exactly right, which is OK. I guess that other script just doesn't store information in the generated header. This script failing might actually be catastrophic.

lindsayad commented on framework/src/base/MooseApp.C:
```diff
@@ -1887,17 +1873,20 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
   //     1) make sure we haven't already loaded this library
   // AND 2) make sure we have a library name (we won't for static linkage)
-  auto dyn_lib_it = _lib_handles.find(library_filename);
+  //
```
```suggestion
```

lindsayad commented on framework/src/base/MooseApp.C:
```diff
@@ -1887,17 +1873,20 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
-  auto dyn_lib_it = _lib_handles.find(library_filename);
+  //
+  // Note: Here was are going to assume uniqueness based on the filename alone. This has significant
+  // implications for applications that have have "diamond" inheritance of libraries (usually
```
```suggestion
  // implications for applications that have "diamond" inheritance of libraries (usually
```

lindsayad commented on framework/src/base/MooseApp.C:
```diff
@@ -1887,17 +1873,20 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
+  //
+  // Note: Here was are going to assume uniqueness based on the filename alone. This has significant
+  // implications for applications that have have "diamond" inheritance of libraries (usually
+  // modules) We will only load one of those libraries, versions be damned.
```
```suggestion
  // modules). We will only load one of those libraries, versions be damned.
```

lindsayad commented on framework/src/base/MooseApp.C:
```diff
@@ -1887,17 +1873,20 @@ MooseApp::loadLibraryAndDependencies(const std::string & library_filename,
-
     // Assemble the actual filename using the base path of the *.la file and the dl_lib_filename
-    std::string dl_lib_full_path = lib_name_parts.first + '/' + dl_lib_filename;
+    const auto & dl_lib_full_path = dir + '/' + dl_lib_filename;
```
```suggestion
    const auto dl_lib_full_path = dir + '/' + dl_lib_filename;
```

lindsayad commented on framework/src/base/MooseApp.C:
```diff
@@ -1997,6 +1988,32 @@ MooseApp::getLoadedLibraryPaths() const
+{
+  std::set<std::string> paths;
+
+  if (library_path != "")
```
```suggestion
  if (!library_path.empty())
```

permcody committed:
Template the Modules::registerAll method

permcody committed:
Template ModulesApp::registerAll

We need to template the modules registration method since when we dlopen open an
application we can wind up with multiple module loaders in memory each with that
method. Those methods however won't have the same information for different apps
due to all the conditional ifdefs for enabling different modules. By templating
we create unique entry points so we can find the right version corresponding to
our application. We'll keep the old method around and deprecate it.

permcody committed:
Cleanup Makefile a bit

permcody committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

permcody committed:
Don't load dependencies automatically

lindsayad commented:
So we really can't do better than the 3.85% coverage rate?

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -79,6 +79,10 @@ MultiApp::validParams()
+  params.addParam<bool>("library_load_dependencies",
+                        false,
+                        "Tells MOOSE to manually load library dependencies. This should not be "
+                        "necessary and is hear for debugging/troubhleshooting.");
```
```suggestion
                        "necessary and is hear for debugging/troubleshooting.");
```

lindsayad commented on modules/module_loader/include/ModulesApp.h:
```diff
@@ -21,7 +100,131 @@ class ModulesApp : public MooseApp
   static void registerExecFlags(Factory & factory);
 };
+
+///@{
+/**
+ * Dummy methods to clear unused parameter warnings from compiler
+ */
+void
+clearUnusedWarnings(Factory & /*factory*/)
+{
+}
+void
+clearUnusedWarnings(Syntax & /*syntax*/, ActionFactory & /*action_factory*/)
+{
+}
+///@}
+
```
```suggestion
```

permcody replied:
I don't understand why these are here. I didn't put them in. They are called in the current registration methods. They've always been empty. Maybe something Robert was doing?

lindsayad commented on modules/module_loader/include/ModulesApp.h:
```diff
@@ -21,7 +100,131 @@ class ModulesApp : public MooseApp
+  ExternalPetscSolverApp::registerAll(f, af, s);
+#endif
+
+  clearUnusedWarnings(f);
+  clearUnusedWarnings(s, af);
```
```suggestion
  libmesh_ignore(f, s, af);
```

permcody replied:
Any thoughts on why these are here in the first place? It was a I don't understand what these are for so don't touch them, "inaction".

lindsayad replied:
If no modules are defined, then these variables would be unused

loganharbour commented:
> So we really can't do better than the 3.85% coverage rate?

I can try to add installed runs to coverage. But I really have _no_ idea how that'd work.

permcody commented:
> > So we really can't do better than the 3.85% coverage rate?
> 
> I can try to add installed runs to coverage. But I really have _no_ idea how that'd work.

Looking at the coverage report now. So the `load_dependencies` branch is always false now (that is the piece that will be used for debugging only, but should be removable). The new `getLibrarySearchPaths` method only gets called on failure, which we admittedly don't have. The one thing that I don't understand is why some of the other lines aren't covered (like the structured binding). That's right in the base implementation. I'm guessing we only do coverage on the main module run and not on "test subs"? Note that the dynamic loading tests only run on the latter. 

Testing for this capability is bad no matter what in lieu of something new (like I described on the PR). If we want testing that actually matters, we're going to need to engineer some real tests with some of the nuclear herd. I have ideas on that. I'll start the dialog with Logan and Jason in the cube tomorrow. I would suggest that we don't hold this up too much though in lieu of testing. I don't have much assurance that any of the existing tests are good enough to test what we need to test.

permcody commented:
In lieu of testing, my plan will be to work with Yinbin (on the Microreactor apps) and April (on Cardinal) as soon as this is merged. Yinbin is already working with dynamic loading and updating his apps to use it by default. Ideally, they can quite compling their own "mega app. April is trying to do the same. They are compiling Griffin into Cardinal, but she's very interested on using this capability. For her, the current code is failing so she's still compiling into Cardinal. With this PR merged, I want to give it another shot.

permcody commented on framework/include/base/MooseApp.h:
```diff
@@ -964,7 +970,9 @@ class MooseApp : public ConsoleStreamInterface,
-  void loadLibraryAndDependencies(const std::string & library_filename, const Parameters & params);
+  void loadLibraryAndDependencies(const std::string & library_filename,
+                                  const Parameters & params,
+                                  bool load_dependencies = true);
```
It strictly doesn't, but I plan to pull this out sooner than later. The whole recursive routine can hopefully go! I'm not kidding when I say I expect 70% of the existing lines to go. The new logic should be to load the named module given the library paths and fail when not found, period. No opening and parsing library archives, no recursive loading.

permcody committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad committed:
Use libmesh_ignore on potentially unused variables

permcody commented:
Thanks Alex!

loganharbour commented:
We're more or less ready to test this at scale with all of our codes with the container squashing.

permcody commented:
We need to update all the apps (at least the ones we care about) to call
the new templated registration method. Then things should work!

On Mon, May 1, 2023 at 7:18 PM Logan Harbour ***@***.***>
wrote:

> We're more or less ready to test this at scale with all of our codes with
> the container squashing.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/24064#issuecomment-1530742135>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOICPPEMRBN6KQBVPFHTXEBOHNANCNFSM6AAAAAAW5VUFFM>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>
